cgen: (void)pointers messing

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
```
OS: macos, macOS, 12.4, 21F79
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 13.1.6 (clang-1316.0.21.2.5)

getwd: /Users/hunam/Documents/dev/testing-v
vmodules: /Users/hunam/.vmodules
vroot: /Applications/v
vexe: /Applications/v/v
vexe mtime: 2022-07-09 15:02:55
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.0 946da55.1ae11b4

Git version: git version 2.32.1 (Apple Git-133)
Git vroot status: 0.3-82-g1ae11b41
.git/config present: true
thirdparty/tcc status: thirdparty-macos-arm64 173c526e

```
<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
mut a := voidptr(0)
mut b := 123
a = &b
println(*a)
```

**What did you expect to see?**
I think it should work but not sure though.
 
**What did you see instead?**
```c
/tmp/v_501/c.5320612907336421070.tmp.c:2116:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/c.5320612907336421070.tmp.c:12020:22: error: passing 'void' to parameter of incompatible type 'voidptr' (aka 'void *')
        println(voidptr_str(*a));
                            ^~
/tmp/v_501/c.5320612907336421070.tmp.c:7778:28: note: passing argument to parameter 'nn' here
string voidptr_str(voidptr nn) {
                           ^
1 warning and 1 error generated.
```

--- a/vlib/v/ast/types.v
+++ b/vlib/v/ast/types.v
@@ -332,6 +332,11 @@ pub fn (typ Type) is_pointer() bool {
 }
 
 [inline]
+pub fn (typ Type) is_voidptr() bool {
+	return typ.idx() == ast.voidptr_type_idx
+}
+
+[inline]
 pub fn (typ Type) is_real_pointer() bool {
 	return typ.is_ptr() || typ.is_pointer()
 }

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3305,6 +3305,9 @@ pub fn (mut c Checker) prefix_expr(mut node ast.PrefixExpr) ast.Type {
 			s := c.table.type_to_str(right_type)
 			c.error('invalid indirect of `$s`', node.pos)
 		}
+		if right_type.is_voidptr() {
+			c.error('cannot dereference to void', node.pos)
+		}
 	}
 	if node.op == .bit_not && !right_type.is_int() && !c.pref.translated && !c.file.is_translated {
 		c.error('operator ~ only defined on int types', node.pos)

--- a/vlib/v/checker/tests/voidptr_dereference_err.out
+++ b/vlib/v/checker/tests/voidptr_dereference_err.out
@@ -1,0 +1,6 @@
+vlib/v/checker/tests/voidptr_dereference_err.vv:5:10: error: cannot dereference to void
+    3 |     mut b := 123
+    4 |     a = &b
+    5 |     println(*a)
+      |             ^
+    6 | }

--- a/vlib/v/checker/tests/voidptr_dereference_err.vv
+++ b/vlib/v/checker/tests/voidptr_dereference_err.vv
@@ -1,0 +1,6 @@
+fn main() {
+	mut a := voidptr(0)
+	mut b := 123
+	a = &b
+	println(*a)
+}

